targets/TARGET_STM/stm_spi_api.c

```diff
@@ -1364,10 +1375,15 @@ void spi_abort_asynch(spi_t *obj)
     NVIC_DisableIRQ(irq_n);
 
     // clean-up
-    __HAL_SPI_DISABLE(handle);
+    LL_SPI_Disable(SPI_INST(obj));
```

Comment:
How did you come up with this fix ? Is spi_abort_asynch being used in your tests ? 
or did you plan to modify init_spi function ?

Reply:
> Is spi_abort_asynch being used in your tests ?

Yes, I have updated demo project to invoke `SPI::abort_all_transfers` between sensor API calls. It's invoked when no actual transfers ongoing, but it shouldn't have any side effects. Although my test shows that this call causes data corruption of the subsequent transmissions, so I found and fixed the error in the `spi_abort_asynch`. Demo project fragment:

https://github.com/vznncv/mbed-os-stm32-spi-3-wire-demo/blob/acadb2e2c213b2b986fec8ac6362e2dbd16749c3/src/main.cpp#L1002-L1008

```.cpp
            // call SPI:spi_abort_all_transfers to check that it has no side effects
            api->call_spi_abort_all_transfers();

            // transmit data to BMX160
            err_write = api->register_write(base_reg, out_buf, buf_len);
            // receive data from BMX160
            err_read = api->register_read(base_reg, in_buf, buf_len);
```

https://github.com/vznncv/mbed-os-stm32-spi-3-wire-demo/blob/acadb2e2c213b2b986fec8ac6362e2dbd16749c3/src/main.cpp#L778-L781

```.cpp
    void call_spi_abort_all_transfers()
    {
        return _spi.abort_all_transfers();
    }
```

> did you plan to modify init_spi function ?

Unlike `spi_abort_asynch` the `init_spi` function already has correct code that doesn't enable SPI in 3 wire mode:

https://github.com/ARMmbed/mbed-os/blob/d2a88f44005eb61b0f1692c4d26601a1d6c68461/targets/TARGET_STM/stm_spi_api.c#L139-L146

The only one thing that I have added to `init_spi` - explicit RX buffer/register cleanup:

```.cpp
    /* In some cases after SPI object re-creation SPI overrun flag may not
     * be cleared, so clear RX data explicitly to prevent any transmissions errors */
    spi_flush_rx(obj);
```

The necessity of these lines I also discovered by my test project (https://github.com/vznncv/mbed-os-stm32-spi-3-wire-demo/blob/40d1556b049a46e82fc2ed214fd4d5340e1e6814/src/main.cpp). The demo re-creates `SPI` object for each test run with different parameters (spi frequency, api type). So If previous test run fails, it may left some data in the RX buffer/register that causes failure of subsequent test runs.

Reply:
ok clear !